src/zarr/core/metadata/dtype.py

```diff
@@ -0,0 +1,576 @@
+    return float(data)
+
+
+def float_to_json_v3(data: float | np.floating[Any]) -> JSONFloat:
+    # v3 can in principle handle distinct NaN values, but numpy does not represent these explicitly
+    # so we just re-use the v2 routine here
+    return float_to_json_v2(data)
+
+
+def float_to_json(data: float | np.floating[Any], zarr_format: ZarrFormat) -> JSONFloat:
+    """
+    convert a float to JSON as per the zarr v3 spec
+    """
+    if zarr_format == 2:
+        return float_to_json_v2(data)
+    else:
+        return float_to_json_v3(data)
+    raise ValueError(f"Invalid zarr format: {zarr_format}. Expected 2 or 3.")
+
+
+def complex_to_json_v2(data: complex | np.complexfloating) -> JSONFloat:
+    return float_to_json_v2(data)
+
+
+def complex_to_json_v3(data: complex | np.complexfloating) -> tuple[JSONFloat, JSONFloat]:
+    return float_to_json_v3(data.real), float_to_json_v3(data.imag)
+
+
+def complex_to_json(
+    data: complex | np.complexfloating, zarr_format: ZarrFormat
+) -> tuple[JSONFloat, JSONFloat] | JSONFloat:
+    if zarr_format == 2:
+        return complex_to_json_v2(data)
+    else:
+        return complex_to_json_v3(data)
+    raise ValueError(f"Invalid zarr format: {zarr_format}. Expected 2 or 3.")
+
+
+def float_from_json_v2(data: JSONFloat) -> float:
+    match data:
+        case "NaN":
+            return float("nan")
+        case "Infinity":
+            return float("inf")
+        case "-Infinity":
+            return float("-inf")
+        case _:
+            return float(data)
+
+
+def float_from_json_v3(data: JSONFloat) -> float:
+    # todo: support the v3-specific NaN handling
+    return float_from_json_v2(data)
+
+
+def float_from_json(data: JSONFloat, zarr_format: ZarrFormat) -> float:
+    if zarr_format == 2:
+        return float_from_json_v2(data)
+    else:
+        return float_from_json_v3(data)
+
+
+def complex_from_json_v2(data: JSONFloat, dtype: Any) -> np.complexfloating:
+    return dtype.type(data)
+
+
+def complex_from_json_v3(data: tuple[JSONFloat, JSONFloat], dtype: Any) -> np.complexfloating:
+    return dtype.type(complex(*data))
+
+
+def complex_from_json(
+    data: tuple[JSONFloat, JSONFloat], dtype: Any, zarr_format: ZarrFormat
+) -> np.complexfloating:
+    if zarr_format == 2:
+        return complex_from_json_v2(data, dtype)
+    else:
+        if check_json_complex_float_v3(data):
+            return complex_from_json_v3(data, dtype)
+        else:
+            raise TypeError(f"Invalid type: {data}. Expected a sequence of two numbers.")
+    raise ValueError(f"Invalid zarr format: {zarr_format}. Expected 2 or 3.")
+
+
+TDType = TypeVar("TDType", bound=np.dtype[Any])
+TScalar = TypeVar("TScalar", bound=np.generic)
+
+
+class DTypeWrapper(Generic[TDType, TScalar], ABC, Metadata):
+    name: ClassVar[str]
+    dtype_cls: ClassVar[type[TDType]]  # this class will create a numpy dtype
+    kind: ClassVar[DataTypeFlavor]
+    default_value: TScalar
```

Comment:
child classes define a string `name` (which feeds into the zarr metadata), a dtype class `dtype_cls` (which gets assigned automatically from the generic type parameter) , a string `kind` (we use this for classifying scalars internally), and a default value (putting this here seems simpler than maintaining a function that matches dtype to default value, but we could potentially do that)